src/relations/index.ts

```diff
@@ -2,22 +2,36 @@ import { Express } from 'express';
 import mongoose, { CallbackError } from 'mongoose';
 import { authorizeRequest } from '../auth';
 import { relationSchema } from './schema';
+import { fixId } from '../utils';
 
 const Relation = mongoose.model('relations', relationSchema);
 
 export default async function(app: Express) {
 
-  app.get('/api/v1/relations', (_, res) => {
-    Relation.find().distinct('_id', function (err: Error, relations: string[]) {
+  /**
+   * Get relations for a resource.
+   *
+   * Query params:
+   * subjectId | objectId | id : resourceId
+   */
+  app.get('/api/v1/relations', (req, res) => {
+    let resourceId = req.query.subjectId || req.query.objectId || req.query.id
```

Comment:
If the client can send any of these fields, it seems to me we should be performing different types of requests based on which fields are actually being sent. And there should still be a way of getting a specific relation by its ID.

Reply:
Perhaps you can help me out here as it was confusing to me what exactly should be going on. At first I assumed it should work as you just described it. But the problem is that there are only two spots in the code that call `ResourceController.getRelations` and neither of them use subjectId or objectId. They just use id which should logically correspond to the relation Id in mongo but it does not. They are using `id=resourceId` where they should be using two requests, one with `subjectId=resourceId` and the other with `objectId=resourceId`. Since this code was working with the previous resource library, I assume it was doing something similar to my implementation where it searches for the resourceId in any of the fields. In which case I suppose I could also search the relation id field in this function.

We could implement it the way you said, which I agree makes sense, but we would have to make some changes to the ayamel backend.

I'd need some clarification on the distinction between subject and object Id before I could make those changes. My understanding is that subjectId is the related object and objectId is the actual object. Translating that to ayamel terms, the object is the video and the subject is the caption file? It feels backwards but I think that's the behavior I observed in the resource library mongo database. Also, can a resource be a subject and an object in different relations? Are there any limitations as to what can be a subject or object? In the scope of Ayamel as it stands can there be cross over where a video is a subject in one relation and an object in another?

Reply:
```there are only two spots in the code that call ResourceController.getRelations and neither of them use subjectId or objectId. They just use id which should logically correspond to the relation Id in mongo but it does not.```

In that case, what we do on this end with `subjectId` and `objectId` fields should be irrelevant--we can have plain `id` trigger the "subjectId *or* objectId" codepath, and also treat `subjectId` and `objectId` separately.

```My understanding is that subjectId is the related object and objectId is the actual object. Translating that to ayamel terms, the object is the video and the subject is the caption file?```

Which resource is the subject and which is the object depends on the semantics of the specific relation type. Captions are connected to videos with a "transcript_of" relation type--so the caption resource "is the transcript of" the video resource, which makes the caption resource the subject and the video resource the object.

```Also, can a resource be a subject and an object in different relations?```
Yes.

```Are there any limitations as to what can be a subject or object?```

No.

```In the scope of Ayamel as it stands can there be cross over where a video is a subject in one relation and an object in another?```

Yes.

Reply:
@gliese1337 I made the changes. There is another issue that should also be fixed that is causing the updated resoureces to lose their language codes. I think it's because we update it using `findOneAndUpdate` which does not trigger the pre save language hook that we defined. https://mongoosejs.com/docs/middleware.html#notes I tried to get 
```
schema.pre('findOneAndUpdate', ...)
schema.pre('updateOne', ...)
```
to work but for some reason the hooks don't run. I'll give this another shot. Just thought it might be worth asking you if you know how to handle this.

Reply:
I just pushed another change to the search logic; it is very rare (maybe impossible?) that the same resource would be both subject and object for a single relation, so it doesn't seem very useful to special-case that, compared to the case where you want to get relations that have a resource a either subject or object.

Do we know exactly where the language codes are being lost? Is it when files are set, or when we `PUT` a new version of a resource directly?

Reply:
The language codes are lost om subsequent saves. The pre logic to set a default is good. The problem is when we update the resource using findoneandupdate. The save hook doesn't run in that case.

Reply:
So that's on `PUT`s *and* file sets, then?

It seems to me that the hooks not being run, while independently annoying, should not cause this specific issue. Once language codes are validated once, the data should not change when updating other fields, regardless of whether or not the `save` hook runs.

Reply:
It only happens with the put endpoint because we update the entire resource. 

When we set the files, we only update that part of the resource, so this isn't a problem.

Reply:
Aha! I would think that running the save hooks wouldn't fix the issue anyway, then, since it would still overwrite the language fields. Are we not `PUT`ting updated documents with the language fields still set?

If we need to do manual document merges, that'll get complicated... but if getting the save hooks to run really is the fix, and we want a `PUT` to just do a whole-sale document replacement without data merge, I think I can get that in tonight.

Reply:
So the use case here is that I create a caption track in captionaider, click save, make some more modifications and then press save again. Since the front-end doesn't see the updated resource with the 'zxx' lang code substituted for '' (no lang code), it sends the put request without a lang code. I guess making sure the lang code validation runs is good and will fix the issue but we could also just update the frontend to always send a lang code or something. Maybe doing both would be good. 🤷

Reply:
OK, I just pushed something that *should* fix `PUT`s so they run the `save` validation hook.